core/include/vecmem/utils/impl/copy.ipp

```diff
@@ -108,7 +127,18 @@ void copy::setup(data::jagged_vector_buffer<TYPE>& data) {
     // "Set up" the inner vector descriptors, using the host-accessible data.
     for (typename data::jagged_vector_buffer<TYPE>::size_type i = 0;
          i < data.m_size; ++i) {
-        setup(data.host_ptr()[i]);
+        // Find the first "inner vector" that has a non-zero capacity, and is
+        // resizable.
+        if ((data.host_ptr()[i].capacity() != 0) &&
+            (data.host_ptr()[i].size_ptr() != nullptr)) {
+            // Initialise the size values using this "inner vector's" pointer.
+            do_memset(sizeof(typename data::vector_buffer<TYPE>::size_type) *
+                          (data.m_size - i),
+                      data.host_ptr()[i].size_ptr(), 0);
```

Comment:
So this way it just sets the entire size block to 0 with a single memset, right?

Reply:
Yes.

I may have overcomplicated the code here a bit, but I wasn't entirely sure whether empty (zero capacity) inner vectors would have a valid "size pointer" or not. Since we do need to check for that in `vecmem::copy::get_size(...)`. So to be safe, I added this for-loop to find the first inner vector that would have a valid "size pointer", and then perform the initialisation of the sizes based on that.

Reply:
Hm, I'm not sure I follow yet. Is this because the size of the size block is implicitly given as the location of the first non-empty vector? Or why to you need the vector's size pointers for this?

Reply:
In the current implementation resizable (jagged) vectors all start from zero sizes. Which is very misleading, I very much agree. Since the UI of these classes suggests that we can create the buffers with a non-zero size, and a capacity larger/equal to that size. But in fact, at the moment the starting sizes of resizable (jagged) vectors are always zero. (We can talk about the reasons for this in more detail in let's say an "issue".)

Since the resizable (jagged) vectors are assumed to be empty at the start, we really only need to make sure that in the allocated (device) memory the "size variables" would be initialised to zero. It doesn't matter how the memory which will hold the payload, is initialised. Since we'll write some values explicitly to that area as we fill the vector.

`size_ptr()` is what returns the pointer to the size variable in "global memory" on the device. This is what we can use to read/write the variable from the host code, and this is also what the device code uses to atomically interact with that variable.

Previously the code didn't assume anything about the internal memory layout of `vecmem::data::jagged_vector_buffer`. It initialised the sizes of the inner vectors one-by-one. Leading to a tremendous number of `memset(...)` operations, as @konradkusiak97 observed. Now I decided to hardcode knowledge into `vecmem::copy` about what the memory layout of the buffer would be. With this slightly non-trivial code... 😛